theme/svg.go

```diff
@@ -0,0 +1,111 @@
+package theme
+
+import (
+	"encoding/hex"
+	"encoding/xml"
+	"fmt"
+	"image/color"
+	"io"
+	"io/ioutil"
+)
+
+// SVG holds the unmarshaled XML from a Scalable Vector Graphic
+type SVG struct {
+	XMLName  xml.Name      `xml:"svg"`
+	XMLNS    string        `xml:"xmlns,attr"`
+	Width    string        `xml:"width,attr"`
+	Height   string        `xml:"height,attr"`
+	ViewBox  string        `xml:"viewBox,attr"`
+	Paths    []*pathObj    `xml:"path"`
+	Rects    []*rectObj    `xml:"rect"`
+	Polygons []*polygonObj `xml:"polygon"`
+	Groups   []*group      `xml:"g"`
+}
+
+type pathObj struct {
+	XMLName xml.Name `xml:"path"`
+	Fill    string   `xml:"fill,attr"`
+	D       string   `xml:"d,attr"`
+}
+
+type rectObj struct {
+	XMLName xml.Name `xml:"rect"`
+	Fill    string   `xml:"fill,attr"`
+	X       string   `xml:"x,attr"`
+	Y       string   `xml:"y,attr"`
+	Width   string   `xml:"width,attr"`
+	Height  string   `xml:"height,attr"`
+}
+
+type polygonObj struct {
+	XMLName xml.Name `xml:"polygon"`
+	Fill    string   `xml:"fill,attr"`
+	Points  string   `xml:"points,attr"`
+}
+
+type group struct {
+	XMLName  xml.Name      `xml:"g"`
+	ID       string        `xml:"id,attr"`
+	Paths    []*pathObj    `xml:"path"`
+	Rects    []*rectObj    `xml:"rect"`
+	Polygons []*polygonObj `xml:"polygon"`
+}
+
+func replacePathsFill(paths []*pathObj, hexColor string) {
+	for _, path := range paths {
+		if path.Fill != "none" {
+			path.Fill = hexColor
+		}
+	}
+}
+
+func replaceRectsFill(rects []*rectObj, hexColor string) {
+	for _, rect := range rects {
+		if rect.Fill != "none" {
+			rect.Fill = hexColor
+		}
+	}
+}
+
+func replacePolygonsFill(polys []*polygonObj, hexColor string) {
+	for _, poly := range polys {
+		if poly.Fill != "none" {
+			poly.Fill = hexColor
+		}
+	}
+}
+
+func replaceGroupObjectFill(groups []*group, hexColor string) {
+	for _, grp := range groups {
+		replacePathsFill(grp.Paths, hexColor)
+		replaceRectsFill(grp.Rects, hexColor)
+		replacePolygonsFill(grp.Polygons, hexColor)
+	}
+}
+
+// ReplaceFillColor alters an SVG objects fill color.  Note that if an SVG with multiple fill
+// colors is being operated upon, all fills will be converted to a single color.  Mostly used
+// to recolor Icons to match the theme's IconColor.
+func (s *SVG) ReplaceFillColor(reader io.Reader, color color.Color) error {
+	bSlice, err := ioutil.ReadAll(reader)
+	if err != nil {
+		return err
+	}
+
+	if err := xml.Unmarshal(bSlice, &s); err != nil {
```

Comment:
I think that this should be part of a separate function - something like svgFromXML?
This essentially loads the SVG struct and so should not really be inside a method that performs a transformation on it.

That way the error from the colour change is not needed.

Reply:
Added the function.